### ISSUE TITLE
Change initElements parameter from WebDriver to SearchContext

### ISSUE THREAD

**Issue by Jiwari** [SeleniumHQ/selenium#9549]
## 🚀 Feature Proposal

Update `PageFactory` methods signature from `WebDriver` to `SearchContext`. 

## Motivation

Allow elements to be passed as an argument to `PageFactory#initElements` methods and be used as the `SearchContext`.  

My objective here is to solve issues related to shadow doms and elements lookup.

I currently have this changed implemented on my project which has a lot of shadow DOMs on the page, and being able to provide the shadow element to `initElements` allows me to use `PageFactory` with the shadow element as the search context, providing WebDriver would never work as the elements are "hidden" inside the shadow DOM.  

## Example

Only 3 methods need to be updated.

Currently:
```java
  public static void initElements(WebDriver driver, Object page) {
    initElements(new DefaultElementLocatorFactory(driver), page);
  }

  public static <T> T initElements(WebDriver driver, Class<T> pageClassToProxy) {
    T page = instantiatePage(driver, pageClassToProxy);
    initElements(driver, page);
    return page;
  }

  private static <T> T instantiatePage(WebDriver driver, Class<T> pageClassToProxy) {
    try {
      try {
        Constructor<T> constructor = pageClassToProxy.getConstructor(WebDriver.class);
        return constructor.newInstance(driver);
      } catch (NoSuchMethodException e) {
        return pageClassToProxy.getDeclaredConstructor().newInstance();
      }
    } catch (ReflectiveOperationException e) {
      throw new RuntimeException(e);
    }
  }
```

With changes:
```java
  public static void initElements(SearchContext searchContext, Object page) {
    initElements(new DefaultElementLocatorFactory(searchContext), page);
  }

  public static <T> T initElements(SearchContextsearchContext, Class<T> pageClassToProxy) {
    T page = instantiatePage(searchContext, pageClassToProxy);
    initElements(searchContext, page);
    return page;
  }

  private static <T> T instantiatePage(SearchContext searchContext, Class<T> pageClassToProxy) {
    try {
      try {
        Constructor<T> constructor = pageClassToProxy.getConstructor(WebDriver.class);
        return constructor.newInstance(searchContext);
      } catch (NoSuchMethodException e) {
        return pageClassToProxy.getDeclaredConstructor().newInstance();
      }
    } catch (ReflectiveOperationException e) {
      throw new RuntimeException(e);
    }
  }
```

No further changes would be needed as upstream classes already use `SearchContext` as type instead of `WebDriver`

My current concern is about `instantiatePage` change. Currently Selenium tries to instantiate the page with 2 different constructors (`WebDriver` arg and parameterless), I am wondering if there should be a third for `SearchContext` between the two existing ones, and if it would be useful.

**Comment by Jiwari** [comment]
I will work on a branch with these changes later this weekend

**Comment by diemol** [comment]
I would recommend to avoid creating a PR before having a chat on the approach, otherwise the PR might get rejected if it does not align with the implementation idea that Java maintainers have (whatever the idea is). Shadow DOM landed in the W3C spec a while ago and we are waiting for browser vendors to support it (also, follow #5869). In the meantime, the bindings maintainers are starting to implement the endpoints for it.

If you'd like to contribute and have a chat, please join us in our [Slack/IRC channel](https://www.selenium.dev/support/), maintainers normally hangout in the `#selenium-tlc` channel, and probably you can discuss your approach with @shs96c and others.

**Comment by diemol** [comment]
Closed via #9557
